File: packages/application/src/ports/vector-store.ts
export interface VectorDoc {
  id: string;
  vector: number[];
  payload: Record<string, unknown>;
}

export interface SearchResult {
  id: string;
  score: number;
  payload: Record<string, unknown>;
}

export interface VectorStorePort {
  upsert(collection: string, docs: VectorDoc[]): Promise<void>;
  search(
    collection: string,
    vector: number[],
    topK: number,
    filter?: Record<string, unknown>,
  ): Promise<SearchResult[]>;
  delete(collection: string, ids: string[]): Promise<void>;
}

File: services/intelligence/app/ports/vector_store.py
from dataclasses import dataclass
from typing import Any, Protocol


@dataclass(slots=True, frozen=True)
class VectorDoc:
    id: str
    vector: list[float]
    payload: dict[str, Any]


@dataclass(slots=True, frozen=True)
class SearchResult:
    id: str
    score: float
    payload: dict[str, Any]


class VectorStorePort(Protocol):
    async def upsert(self, collection: str, docs: list[VectorDoc]) -> None: ...

    async def search(
        self,
        collection: str,
        vector: list[float],
        top_k: int,
        filter: dict[str, Any] | None = None,
    ) -> list[SearchResult]: ...

    async def delete(self, collection: str, ids: list[str]) -> None: ...
